### PR TITLE
upload using programmer

### PR DESCRIPTION
**Pull request by czgtest**

### CODE REVIEW

File: src/arduino/programmerManager.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    private _currentprogrammer: ProgrammerList;
+
+    private _programmervalue: string;
+
+    private _programmerStatusBar: vscode.StatusBarItem;
+
+    /**
+     * Need refresh Arduino IDE's setting when starting up.
+     * @param {boolean} force - Whether force initialize the arduino
+     */
```
Comment (Sneezry) [inline src/arduino/programmerManager.ts]: clean up

Comment (czgtest) [reply src/arduino/programmerManager.ts]: Done

File: src/common/constants.ts [hunk 2]
```diff
@@ -1,38 +1,40 @@
...
-
-export const statusBarPriority = {
-    PORT: 20,
-    OPEN_PORT: 30,
-    BAUD_RATE: 40,
-    BOARD: 60,
-    ENDING: 70,
-    SKETCH: 80,
-};
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (Sneezry) [inline src/common/constants.ts]: It seems this file'ending is CRLF in previous version. Let's keep it as CRLF.

Comment (czgtest) [reply src/common/constants.ts]: Done

File: package.json [hunk 3]
```diff
@@ -457,7 +467,7 @@
         "mimetypes": [
           "text/x-code-output"
         ]
-      }, {
+        }, {
```
Comment (Sneezry) [inline package.json]: alignment

File: src/extension.ts [hunk 4]
```diff
@@ -193,6 +193,32 @@ export async function activate(context: vscode.ExtensionContext) {
...
+            status.compile = "upload";
+            try {
+                await ArduinoContext.arduinoApp.programmerManager.selectProgrammer();
+            } catch (ex) {
+            }
+            delete status.compile;
+        }
+    }, () => {
+        return { board: ArduinoContext.boardManager.currentBoard.name };
+     });
```
Comment (Sneezry) [inline src/extension.ts]: alignment

**Pull request merged**